File: ai-service/utils/encryption.py
"""
AES-256-GCM field-level encryption — Pillar 4: Security
Used to encrypt sensitive source IDs and PII in stored alerts
"""
import base64
import os
from config import ENCRYPTION_KEY


def encrypt_field(plaintext: str) -> str:
    if not plaintext:
        return ""
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
        key = ENCRYPTION_KEY.encode()[:32]
        aesgcm = AESGCM(key)
        nonce = os.urandom(12)
        ct = aesgcm.encrypt(nonce, plaintext.encode(), None)
        return base64.b64encode(nonce + ct).decode()
    except Exception:
        return base64.b64encode(plaintext.encode()).decode()


def decrypt_field(ciphertext: str) -> str:
    if not ciphertext:
        return ""
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
        key = ENCRYPTION_KEY.encode()[:32]
        aesgcm = AESGCM(key)
        raw = base64.b64decode(ciphertext)
        nonce, ct = raw[:12], raw[12:]
        return aesgcm.decrypt(nonce, ct, None).decode()
    except Exception:
        try:
            return base64.b64decode(ciphertext).decode()
        except Exception:
            return "[decryption failed]"

File: backend/utils/encryption.js
"use strict";
const CryptoJS = require("crypto-js");

const KEY = process.env.ENCRYPTION_KEY || "fallback_key_change_in_production!";

/**
 * AES-256 encrypt a string value.
 * Used for Pillar 4 — field-level encryption of sensitive data.
 */
const encryptField = (plaintext) => {
  if (!plaintext) return plaintext;
  return CryptoJS.AES.encrypt(String(plaintext), KEY).toString();
};

/**
 * Decrypt an AES-256 encrypted string.
 */
const decryptField = (ciphertext) => {
  if (!ciphertext) return ciphertext;
  const bytes = CryptoJS.AES.decrypt(ciphertext, KEY);
  return bytes.toString(CryptoJS.enc.Utf8);
};

/**
 * Hash a value with SHA-256 (for checksums/integrity, NOT passwords).
 */
const hashValue = (value) => {
  return CryptoJS.SHA256(String(value)).toString(CryptoJS.enc.Hex);
};

module.exports = { encryptField, decryptField, hashValue };